### PR TITLE
feat: move proxy settings page to deployment options

### LINKED ISSUE #8153: Move "Proxy Settings" to Deployment page and add "WorkspaceProxy" label to dropdown
And refresh latencies not close the modal.

### PR DESCRIPTION
**Pull request by Emyrk**
The Workspace proxy settings page was not showing any more information, so it was moved to the "Deployment" page only viewable by admins. More details can be shown there.

Closes https://github.com/coder/coder/issues/8153

# Images

## Regular user (member)

Screenshot from 2023-06-28 09-16-13

## Admin

Screenshot from 2023-06-28 09-16-18

Notice the more accurate badges. Instead of just `Healthy` and `Unhealthy`. Not shown are the error messages that cause those errors. Need to figure out a way to include them with good styling.

Screenshot from 2023-06-28 09-14-50

# Additional Changes

- Typesgenerated: I added functionality for generic slices
- ServerSide rendering: Return `Region` or `WorkspaceProxy` based on the user's perms

# Future work

Each workspace proxy has 0 or more **warnings** and **errors**. We should show these, I just do not know the best way to style it :cry:.

### CODE REVIEW

**Comment by bpmct** [comment]
Does this close #8153?

**Comment by Emyrk** [comment]
> Does this close #8153?

Yes

File: scripts/apitypings/testdata/genericmap/genericmap.ts [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+  readonly bar: string
+}
+
+// From codersdk/genericmap.go
+export interface FooBuzz<R extends Custom> {
+  readonly something: R[]
+}
+
+// From codersdk/genericmap.go
+export type Custom = Foo | Buzz
```
Comment (Emyrk) [inline scripts/apitypings/testdata/genericmap/genericmap.ts:20]: The typesgenerated addition just allows this.

File: scripts/apitypings/testdata/genericmap/genericmap.go [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+// Not yet supported
+//type FooBuzzMap[R Custom] struct {
+//	Something map[string]R `json:"something"`
+//}
+
+// Not yet supported
+//type FooBuzzAnonymousUnion[R Foo | Buzz] struct {
+//	Something []R `json:"something"`
+//}
```
Comment (Emyrk) [inline scripts/apitypings/testdata/genericmap/genericmap.go:28]: I might get to this later in another PR. I do not need this in this PR.

File: scripts/apitypings/main.go [hunk 3]
```diff
@@ -691,7 +696,16 @@ func (g *Generator) typescriptType(ty types.Type) (TypescriptType, error) {
...
+			var genValue = ""
+			if underlying.GenericValue != "" {
+				genValue = underlying.GenericValue + "[]"
+			}
+			return TypescriptType{
+				ValueType:     underlying.ValueType + "[]",
+				GenericValue:  genValue,
+				AboveTypeLine: underlying.AboveTypeLine,
+				GenericTypes:  underlying.GenericTypes,
+			}, nil
```
Comment (Emyrk) [inline scripts/apitypings/main.go:708]: The only reason this did not exist is because we did not need it yet. It's a trivial addition.

**Review by bpmct — COMMENTED** [review]
Can you add the description for workspace proxies in the dropdown? Users may not know what this region does (is it where I provision workspaces, is it my preferred DERP when I ssh, etc?)

> Workspace proxies are used to reduce the latency of web-based connections to your workspaces.

**Comment by Emyrk** [comment]
> Can you add the description for workspace proxies in the dropdown? Users may not know what this region does (is it where I provision workspaces, is it my preferred DERP when I ssh, etc?)
> 
> > Workspace proxies are used to reduce the latency of web-based connections to your workspaces.



> Can you add the description for workspace proxies in the dropdown? Users may not know what this region does (is it where I provision workspaces, is it my preferred DERP when I ssh, etc?)
> 
> > Workspace proxies are used to reduce the latency of web-based connections to your workspaces.

@bpmct I added that to this ticket: https://github.com/coder/coder/issues/8155

This PR has already grown a lot and I am stuck on how to style some dropdown additions.

File: cli/cliui/table_test.go [hunk 4]
```diff
@@ -188,6 +193,31 @@ foo   foo1        foo3              2022-08-02T15:49:10Z
...
+					Name: stringWrapper{
+						str: "Alice",
+					},
+					Age:         25,
+					NotIncluded: "IgnoreMe",
+				},
+			},
+		}
+		out, err := cliui.DisplayTable(inlineIn, "", []string{"name", "age"})
+		log.Println("rendered table:\n" + out)
```
Comment (Kira-Pilot) [inline cli/cliui/table_test.go:216]: Is this log intentional?

Comment (Emyrk) [reply cli/cliui/table_test.go:216]: Yes, it's helpful to build the expected output. Test logs only log in the error case by default (unless you do `go test -v`)

File: site/src/components/DeploySettingsLayout/Badges.tsx [hunk 5]
```diff
@@ -115,6 +134,11 @@ const useStyles = makeStyles((theme) => ({
     backgroundColor: theme.palette.error.dark,
   },
 
+  warnBadge: {
+    border: `1px solid ${theme.palette.warning.light}`,
+    backgroundColor: theme.palette.warning.dark,
+  },
```
Comment (Kira-Pilot) [inline site/src/components/DeploySettingsLayout/Badges.tsx:140]: Nice! We do now support [writing styled components](https://emotion.sh/docs/styled) as opposed to classes, which we may want to lean into so we can get rid of `makeStyles` eventually. Not something that needs to be changed here - just a heads up.

File: site/src/contexts/ProxyContext.tsx [hunk 6]
```diff
@@ -103,20 +109,39 @@ export const ProxyProvider: FC<PropsWithChildren> = ({ children }) => {
...
+    return resp.regions
+  }
+
+  if (permissions.editWorkspaceProxies) {
+    // Admins should query the more detailed endpoint.
+    query = async (): Promise<Region[]> => {
+      const resp = await getWorkspaceProxies()
+      return resp.regions
+    }
+  }
```
Comment (Kira-Pilot) [inline site/src/contexts/ProxyContext.tsx:135]:
```suggestion
  const query = async (): Promise<Region[]> => {
    const endpoint = permissions.editWorkspaceProxies ? getWorkspaceProxies : getWorkspaceProxyRegions
    const resp = await endpoint()
    return resp.regions
  }
```

**Review by Kira-Pilot — APPROVED** [review]
Pulled down - seems to be working smoothly!

File: site/src/components/DeploySettingsLayout/Badges.tsx (hunk 5 above)
Comment (Emyrk) [reply site/src/components/DeploySettingsLayout/Badges.tsx:140]: Huh TIL. Will have to see some examples in the future.

File: site/src/contexts/ProxyContext.tsx (hunk 6 above)
Comment (Emyrk) [reply site/src/contexts/ProxyContext.tsx:135]: There is always a smaller way :smile:

**Pull request merged**